# Taller 2

Andrés Estupiñán Moncada 

Brayan Andrés Castellanos Pinilla 

Primero, cargamos las librerias que más usamos 

In [181]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import linprog
import warnings

# Punto 1

In [201]:
warnings.filterwarnings("ignore")
# Objetivo a maximizar 
c = np.array([3, 5])       # La ganancia por unidad de A y de B. (A  y B son los productos)

# Restricciones de desigualdad
A = np.array([  [1, 1],    # Materia prima en unidades por producto
                [2, 4],    # Mano de  obra en unidades por producto
                [1, 2],    # Capital en unidades por producto
                [-1,0],    # Demanda del producto A
                [0,-1]])   # Demanda del producto B
b = np.array([500,1000,2000,-100,-150]) # Restricción 

# Solución del problema
sol = linprog(-c, A_ub=A, b_ub=b, method='simplex') 
print(round(pd.DataFrame(sol.x),2))
sol.fun                    # Encontramos "fun" que es la solución de la maximización y "x" que es los parametros A y B optimizados.

       0
0  200.0
1  150.0


-1350.0

La asignación óptima de A  es de 200 unidades y de B es de 150 unidades y las ganancias de la empresa serían $1350.

# Punto 2

In [204]:
# Objectivo a maximizar 
c2 = np.array([0.1, 0.15, 0.16, 0.08])    # Los retornos de cada fondo

# Restricciones de desigualdad
A2 = np.array([[0.5,    0.3, 0.25, 0.6],  # Valoraciones de las acciones en porcentaje
                  [0.3, 0.1, 0.4 , 0.2],  # Tamaño de las acciones en porcentaje
                  [0.2, 0.6, 0.35, 0.2]]) # Liquidez de las acciones en porcentaje
b2 = np.array([35, 30, 15])               # Restricciones 

# Restricciones de igualdad
A3 = np.array([[1  , 1  , 1   , 1  ]])    # Variables que identifican lo que se invierte en cada fondo
b3 = np.array([100])                      # Capital $100 millones 
# Solución del problema
res2 = linprog(-c2, A_ub=-A2, b_ub=-b2, A_eq=A3, b_eq=b3,method='revised simplex')
print(round(pd.DataFrame(res2.x),2))
round(res2.fun,2)                         # Encontramos "fun" que es la solución de la maximización y "x" que es lo que se invierte en cada fondo optimizado.

       0
0   0.00
1  15.79
2  57.89
3  26.32


-13.74

Una asignación factible sería una invrsión de  0 millones en el fondo 1, 15.79 millones en el fondo 2, 57.89 millones en el fondo 3 y 26.32 millones en el fondo 4. El cual tiene un retorno esperado de 13.74 millones.

# Punto 3

In [205]:
# Objectivo a maximizar 
c6 = np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,1])                 # Maximizar el saldo del ultimo mes 

# Restricciones de igualdad
A6 = np.array([[1,0,0,0,0, 1,0,0, -1,0,0,0,0,0],          
             [-1.01,1,0,0,0, 0,1,0, 1.003,-1,0,0,0,0],
             [0,-1.01,1,0,0, 0,0,1, 0,1.003,-1,0,0,0],
             [0,0,-1.01,1,0, -1.02,0,0, 0,0,1.003,-1,0,0],
             [0,0,0,-1.01,1, 0,-1.02,0, 0,0,0,1.003,-1,0],
             [0,0,0,0,-1.01, 0,0,-1.02, 0,0,0,0,1.003,-1]])  # xj: monto dispuesto a pedir en la línea de crédito en el mes j, para j = 1,2,3,4,5., yj: monto emitido en el papel comercial en el mes j, para j = 1,2,3. y zj : saldo de fondos (FC) en el mes j, para j = 1,2,3,4,5,6.
b6 = np.array([150, 100, -200,200,-50,-300])                 # Restricción Flujo de caja

# Solución del problema
res4 = linprog(-c6, A_eq=A6, b_eq=b6,method='revised simplex')
print(round(pd.DataFrame(res4.x),2))                          # Encontramos "x" que es la desición que se toma con cada variable en cada mes.

         0
0     0.00
1    50.98
2     0.00
3     0.00
4     0.00
5   150.00
6    49.02
7   203.43
8     0.00
9     0.00
10  351.94
11    0.00
12    0.00
13   92.50


-92.5

La recomendación para la empresa es: Para el priimer mes emitir bonos por 150 millones, para el segundo mes pedir un credito de 50.98 millones y emitir un bono por 49.02 millones, para el tercer mes invertir el excedente 351.94 millones y emitir un bono por 203.43 millones. Con todo lo anterior la empresa lograria para el mes 6 tener un excedente de 92.49 millones el cual es el maximo segun las restricciones.  

# Punto 4

In [207]:
# Parametro 
rate = (1+0.0001)**7                 # Rentabilidad neta de los recursos invertidos a corto plazo 

# Objectivo a maximizar 
c7 = np.array([0,0,0,1])             # Maximizar el saldo de la ultima semana   

# Restricciones de desigualdad
A7 = np.array([[-1,0,0,0],           # x: El monto que se invierte y rate es la ganancia de lo invertido en el periodo anterior 
               [rate,-1,0,0],
               [0,rate,-1,0],
               [0,0,rate,-1]])
b7 = np.array([-150, 100, 50,-200])  # Restricción: entradas menos pagos de la empresa 

# Solución del problema
res5 = linprog(-c7, A_ub=-A7, b_ub=-b7,method='simplex')
print(round(pd.DataFrame(res5.x),2)) # Encontramos "x" que es lo que se invierte cada semana y el ultimo valor es la saldo final.

        0
0  150.00
1   50.11
2    0.14
3  200.14


A la empresa se le recomienda para la primera semana invertir 150, para la segunda semana ya despues de haber obtenido las ganancias de la semana 1 se recomienda invertir 50.11, para la tercera semana se le recomienda invertir 0.14. En la semana cuatro tendra un saldo de 200.14 lo que quiere decir que gracias a las inversiones que se hicieron tuvo una ganancia estrategica de 0.14. 

#Punto 5

In [209]:
# Objectivo a maximizar 
c4 = np.array([0, 0, 0, 5,8,10])            # Maximizar las ganacias por departamento

# Restricciones de desigualdad
A4 = np.array([[0,  0,  0,  1  ,0  ,0  ],   # Las primeras variables "X" son las horas por departamento y las segundas variables "Y" son el presupuesto por departamento
               [0,  1,  0,  0  ,0  ,0  ],
               [0,  0,  1,  0  ,0  ,0  ],
               [0,  0,  0,  0  ,1  ,0  ],
               [0,  0,  0,  0  ,0  ,1  ]])
b4 = np.array([2000, 300, 250,2000,2000])   # Restricción de horas y restricción presupuestal por departamento 

# Restricciones de igualdad
A5 = np.array([[1  , 1  , 1   , 0  ,0  ,0  ],[ 0  ,0  ,0, 1  , 1  , 1  ]]) 
b5 = np.array([1000,10000])                 # Restricción total de horas y restricción total presupuestal
# Solución del problema
res3 = linprog(-c4, A_ub=-A4, b_ub=-b4, A_eq=A5, b_eq=b5,method='simplex')
print(round(pd.DataFrame(res3.x),2))       
res3.fun                                    # Encontramos "x" que es lo optimo en horas por departamento y lo optimo en presupuesto por departamento. Luego encontramos "fun" que es la ganancia final de la empresa.

        0
0   450.0
1   300.0
2   250.0
3  2000.0
4  2000.0
5  6000.0


-86000.0

La empresa, para maximizar sus ganancias, debera distribuir los 10.000 millones y la horas de trabajo en la siguiente forma:  En ventas dejar un presupuesto de 2.000 millones y contar con 450 horas de trabajo. En Marketing dejar un presupuesto de 2.000 millones y contar con 300 horas de trabajo. En Producción dejar un presupuesto de 6.000 millones y contar con 250 horas de trabajo. Al final la empresa tentra una ganancia de 86.000 millones. 

#Punto 6

In [179]:
import cvxpy as cp
warnings.filterwarnings("ignore")
# Definimos la matriz de costos de distribución 
cd = np.array([[4,  5,  6,  8  ,10],
               [6,  4,  3,  5  ,8],
               [9,  7,  4,  3  ,4]])
# Definimos matriz de costos fijos
cf = np.array([1000,1000,1000])
# Definimos matriz de producción maxima
pm = np.array([500,500,500])
# Definiomos la variable de desición, si funciona o no la fabrica (es una variable de (si,no))
vd = cp.Variable(3,boolean=True)
# Definimos la matriz de restricción de demanda de cada fabrica 
rd = np.array([80,270,250,160,180])
# Definimos la matriz de incognitas 
x = cp.Variable((3,5), nonneg=True)
# Creamos las restricciones 
restricciones = [] 
for i in range(3): restricciones.append(cp.sum(x[i,:]) <= pm[i]*vd[i])
for j in range(5): restricciones.append(cp.sum(x[:,j]) >= rd[j])
# Creamos la función objetivo 
objetivo = cp.Minimize(cp.sum(cp.multiply(cf,vd))+cp.sum(cp.multiply(cd,x)))
# Solución problema
Problema = cp.Problem(objetivo, restricciones)
Solucion = Problema.solve()
Solucion

5610.0

In [180]:
x.value

array([[  0.,  -0.,  -0.,  -0.,  -0.],
       [ 80., 270., 150.,  -0.,  -0.],
       [ -0.,  -0., 100., 160., 180.]])

La planta que se serrará, con el objetivo de minimizar costos y solo usar lo necesario para cubrir la demanda, es la planta 1. 

La planta dos cubrira la demanda de 80 unidades del punto 1, 270 unidades del punto 2 y 150 unidades del punto 3. La planta tres cubrira la demanda de 100 unidades del punto 3, 160 unidades del punto 4 y 180 unidades del punto 5. Con todo esto, la empresa incurrira en un costo minimo de 5610.   